src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleManager.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Properties;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using NuGet.Versioning;
+
+namespace Microsoft.Azure.WebJobs.Script.BindingExtensionBundle
+{
+    internal class ExtensionBundleManager : IExtensionBundleManager
+    {
+        private readonly IEnvironment _environment;
+        private readonly IOptions<ExtensionBundleOptions> _options;
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+
+        public ExtensionBundleManager(IEnvironment environment,
+                                      IOptions<ExtensionBundleOptions> extensionBundleOptions,
+                                      HttpClient httpClient,
+                                      ILogger<ExtensionBundleManager> logger)
+        {
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _options = extensionBundleOptions ?? throw new ArgumentNullException(nameof(extensionBundleOptions));
+            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public bool TryLocateExtensionBundle(out string bundlePath)
+        {
+            bundlePath = null;
+            string version = null;
+            var paths = new List<string>(_options.Value.ProbingPaths)
+            {
+                _options.Value.DownloadPath
+            };
+
+            for (int i = 0; i < paths.Count; i++)
+            {
+                var path = paths.ElementAt(i);
+                _logger.LogInformation(Resources.LocateExtensionBundle, _options.Value.Id, path);
+                if (FileUtility.DirectoryExists(path))
+                {
+                    var bundleDirectories = FileUtility.EnumerateDirectories(path);
+                    version = FindBestVersionMatch(_options.Value.Version, bundleDirectories);
+                }
+
+                if (!string.IsNullOrEmpty(version))
+                {
+                    bundlePath = Path.Combine(path, version);
+                    _logger.LogInformation(Resources.ExtensionBundleFound);
```

Comment:
Would be nice to log the version as well, so we know which bundle was actually used.